File: mobile/android/core/designsystem/src/main/kotlin/com/slopssaloon/omen/core/designsystem/token/OmenColor.kt
package com.slopssaloon.omen.core.designsystem.token

import androidx.compose.ui.graphics.Color

/**
 * Core semantic color roles. Values are the live `index.css` hexes transcribed in
 * `omen-native-design-system-registry-v1.md` §2.2 — do not hand-edit a hex here without
 * updating the registry row first (registry markdown is the source of truth).
 *
 * Feature code must read [com.slopssaloon.omen.core.designsystem.theme.OmenTheme.color],
 * never construct [OmenColorScheme] or a raw [Color] directly (registry §2.6).
 */
data class OmenColorScheme(
    val bg: Color,
    val surface1: Color,
    val surface2: Color,
    val surface3: Color,
    val border: Color,
    val borderSubtle: Color,
    val textPrimary: Color,
    val textSecondary: Color,
    val textTertiary: Color,
    val accent: Color,
    val accentHover: Color,
    val accentMuted: Color,
    val textOnAccent: Color,
    val omen: Color,
    val umber: Color,
    /** Derived from [accent]; solid stroke color for the focus-ring outline (registry §2.2, §4). */
    val focusRing: Color,
    /** Soft halo layered behind [focusRing] — the "accent @ 40%" reading of the registry row. */
    val focusRingHalo: Color,
    val data: OmenDataSemanticColors,
)

/**
 * Data-semantic invariant tokens (registry §2.3). These never change with a theme pack —
 * risk/data-source/confidence/position/platform meaning must stay fixed across the whole app.
 * Identical values in light and dark by contract, except where the registry documents a
 * light-mode override (risk-low/medium below).
 */
data class OmenDataSemanticColors(
    val riskLow: Color,
    val riskMedium: Color,
    val riskHigh: Color,
    val dataLive: Color,
    val dataStub: Color,
    val dataMock: Color,
    val dataUnavailable: Color,
    val confidenceFloor: Color,
    val confidenceCeiling: Color,
    val posRb: Color,
    val posWr: Color,
    val posQb: Color,
    val posTe: Color,
    val posDef: Color,
    val posK: Color,
    val platformSleeper: Color,
    val platformYahoo: Color,
    val platformEspn: Color,
    val demoText: Color,
    val demoTextSecondary: Color,
)

private val darkDataSemantics = OmenDataSemanticColors(
    riskLow = Color(0xFF34C759),
    riskMedium = Color(0xFFFF9F0A),
    riskHigh = Color(0xFF7E1717),
    dataLive = Color(0xFF34C759),
    dataStub = Color(0xFFFF9F0A),
    dataMock = Color(0xFF636366),
    dataUnavailable = Color(0xFF3A3A3C),
    confidenceFloor = Color(0xFF701020),
    confidenceCeiling = Color(0xFF206F3A),
    posRb = Color(0xFF34D399),
    posWr = Color(0xFF60A5FA),
    posQb = Color(0xFFFB923C),
    posTe = Color(0xFFC084FC),
    posDef = Color(0xFFF472B6),
    posK = Color(0xFFA3A3A3),
    platformSleeper = Color(0xFF1FA3E8),
    platformYahoo = Color(0xFF410093),
    platformEspn = Color(0xFFC81E2C),
    demoText = Color(0xFF7DD3FC),
    demoTextSecondary = Color(0xCCBAE6FD),
)

// Registry §2.3: risk-high and the data-source/confidence/position/platform/demo families are
// identical dark/light by contract; only risk-low/risk-medium document a light override.
private val lightDataSemantics = darkDataSemantics.copy(
    riskLow = Color(0xFF16A34A),
    riskMedium = Color(0xFFD97706),
)

val OmenDarkColors = OmenColorScheme(
    bg = Color(0xFF0A0A0B),
    surface1 = Color(0xFF1C1C1E),
    surface2 = Color(0xFF2C2C2E),
    surface3 = Color(0xFF3A3A3C),
    border = Color(0xFF3A3A3C),
    borderSubtle = Color(0xFF2C2C2E),
    textPrimary = Color(0xFFF5F0E8),
    textSecondary = Color(0xFFAEAEB2),
    textTertiary = Color(0xFF6D6D72),
    accent = Color(0xFFA67C2E),
    accentHover = Color(0xFFC49035),
    accentMuted = Color(0xFF3A2A0A),
    textOnAccent = Color(0xFF0A0A0B),
    omen = Color(0xFF2F7D5B),
    umber = Color(0xFF5A3A25),
    focusRing = Color(0xFFA67C2E),
    focusRingHalo = Color(0xFFA67C2E).copy(alpha = 0.4f),
    data = darkDataSemantics,
)

val OmenLightColors = OmenColorScheme(
    bg = Color(0xFFFAFAF9),
    surface1 = Color(0xFFFFFFFF),
    surface2 = Color(0xFFF5F5F4),
    surface3 = Color(0xFFEBEBEA),
    border = Color(0xFFE5E5E3),
    borderSubtle = Color(0xFFF0F0EE),
    textPrimary = Color(0xFF1C1C1E),
    textSecondary = Color(0xFF6B7280),
    textTertiary = Color(0xFF9CA3AF),
    accent = Color(0xFF7A5C1E),
    accentHover = Color(0xFFA67C2E),
    accentMuted = Color(0xFFFEF3C7),
    textOnAccent = Color(0xFFFAFAF9),
    omen = Color(0xFF1A5C3E),
    umber = Color(0xFF5A3A25),
    focusRing = Color(0xFF7A5C1E),
    focusRingHalo = Color(0xFF7A5C1E).copy(alpha = 0.4f),
    data = lightDataSemantics,
)

File: mobile/ios/OmenIOS/OmenIOS/DesignSystem/OmenColor.swift
import SwiftUI
import UIKit

/// Core semantic color roles. Values are the live `index.css` hexes transcribed in
/// `omen-native-design-system-registry-v1.md` §2.2/§2.3 — do not hand-edit a hex here without
/// updating the registry row first (registry markdown is the source of truth).
///
/// Every token is a trait-aware dynamic `Color` (registry §2.6): feature code reads
/// `OmenColor.xxx` directly and gets the correct dark/light value automatically, with no
/// `@Environment(\.colorScheme)` plumbing required at the call site.
enum OmenColor {
    static let bg = dynamic(dark: 0x0A0A0B, light: 0xFAFAF9)
    static let surface1 = dynamic(dark: 0x1C1C1E, light: 0xFFFFFF)
    static let surface2 = dynamic(dark: 0x2C2C2E, light: 0xF5F5F4)
    static let surface3 = dynamic(dark: 0x3A3A3C, light: 0xEBEBEA)
    static let border = dynamic(dark: 0x3A3A3C, light: 0xE5E5E3)
    static let borderSubtle = dynamic(dark: 0x2C2C2E, light: 0xF0F0EE)
    static let textPrimary = dynamic(dark: 0xF5F0E8, light: 0x1C1C1E)
    static let textSecondary = dynamic(dark: 0xAEAEB2, light: 0x6B7280)
    static let textTertiary = dynamic(dark: 0x6D6D72, light: 0x9CA3AF)
    static let accent = dynamic(dark: 0xA67C2E, light: 0x7A5C1E)
    static let accentHover = dynamic(dark: 0xC49035, light: 0xA67C2E)
    static let accentMuted = dynamic(dark: 0x3A2A0A, light: 0xFEF3C7)
    static let textOnAccent = dynamic(dark: 0x0A0A0B, light: 0xFAFAF9)
    static let omen = dynamic(dark: 0x2F7D5B, light: 0x1A5C3E)
    static let umber = dynamic(dark: 0x5A3A25, light: 0x5A3A25)

    /// Derived from `accent`; solid stroke color for the focus-ring outline (registry §2.2, §4).
    static let focusRing = accent
    /// Soft halo layered behind `focusRing` — the "accent @ 40%" reading of the registry row.
    static let focusRingHalo = dynamic(dark: 0xA67C2E, light: 0x7A5C1E, alpha: 0.4)

    enum Data {
        // Data-semantic invariant tokens (registry §2.3) — never theme-overridden, except the
        // documented risk-low/risk-medium light-mode adjustment below.
        static let riskLow = dynamic(dark: 0x34C759, light: 0x16A34A)
        static let riskMedium = dynamic(dark: 0xFF9F0A, light: 0xD97706)
        static let riskHigh = invariant(0x7E1717)
        static let dataLive = invariant(0x34C759)
        static let dataStub = invariant(0xFF9F0A)
        static let dataMock = invariant(0x636366)
        static let dataUnavailable = invariant(0x3A3A3C)
        static let confidenceFloor = invariant(0x701020)
        static let confidenceCeiling = invariant(0x206F3A)
        static let posRb = invariant(0x34D399)
        static let posWr = invariant(0x60A5FA)
        static let posQb = invariant(0xFB923C)
        static let posTe = invariant(0xC084FC)
        static let posDef = invariant(0xF472B6)
        static let posK = invariant(0xA3A3A3)
        static let platformSleeper = invariant(0x1FA3E8)
        static let platformYahoo = invariant(0x410093)
        static let platformEspn = invariant(0xC81E2C)
        static let demoText = invariant(0x7DD3FC)
        static let demoTextSecondary = invariant(0xBAE6FD, alpha: 0.8)
    }

    /// A color with the same value in both dark and light mode (data-semantic invariant rule).
    private static func invariant(_ hex: UInt32, alpha: CGFloat = 1) -> Color {
        dynamic(dark: hex, light: hex, alpha: alpha)
    }

    private static func dynamic(dark: UInt32, light: UInt32, alpha: CGFloat = 1) -> Color {
        Color(UIColor { traits in
            traits.userInterfaceStyle == .dark
                ? UIColor(omenHex: dark, alpha: alpha)
                : UIColor(omenHex: light, alpha: alpha)
        })
    }
}

private extension UIColor {
    convenience init(omenHex hex: UInt32, alpha: CGFloat) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: alpha
        )
    }
}
